### PR TITLE
#SB-16351: Data product changes to read and write to postgres instead of cassandra

### PR DESCRIPTION
**Pull request by utk14**

### CODE REVIEW

File: platform-modules/batch-models/src/main/scala/org/ekstep/analytics/model/DeviceSummaryModel.scala [hunk 1]
```diff
@@ -65,17 +67,33 @@ object DeviceSummaryModel extends IBatchModelTemplate[String, DeviceInput, Devic
...
-          .joinWithCassandraTable[Option[Long]](Constants.DEVICE_KEY_SPACE_NAME, Constants.DEVICE_PROFILE_TABLE).select("first_access")
-          .on(SomeColumns("device_id"))
-        summary.leftOuterJoin(firstAccessFromCassandra)
-          .map{ x =>
-              val firstAccessValue = x._2._2.getOrElse(Option(0L))
-              if(firstAccessValue.getOrElse(0L) != 0)
-                  x._2._1.copy(firstAccess = firstAccessValue.getOrElse(x._2._1.firstAccess))
-              else x._2._1
+        finally {
+            postgresDB.closeConnection
```
Comment (SanthoshVasabhaktula) [inline platform-modules/batch-models/src/main/scala/org/ekstep/analytics/model/DeviceSummaryModel.scala]: Why is the connection closed here?

File: platform-modules/batch-models/src/main/scala/org/ekstep/analytics/model/DeviceSummaryModel.scala [hunk 2]
```diff
@@ -65,17 +67,33 @@ object DeviceSummaryModel extends IBatchModelTemplate[String, DeviceInput, Devic
...
             val dial_failure = dialcodes_events.filter(f => f.edata.size == 0).length
             val content_downloads = raw.filter(f => "INTERACT".equals(f.eid)).length
-            (index, DeviceSummary(index.device_id, index.channel, CommonUtil.roundDouble(total_ts, 2), total_launches, contents_played, unique_contents_played, content_downloads, DialStats(dial_count, dial_success, dial_failure), DtRange(startTimestamp, endTimestamp), syncts, startTimestamp))
+            (index, DeviceSummary(index.device_id, index.channel, CommonUtil.roundDouble(total_ts, 2), total_launches, contents_played, unique_contents_played, content_downloads, DialStats(dial_count, dial_success, dial_failure), DtRange(startTimestamp, endTimestamp), syncts, new Timestamp(startTimestamp)))
+        }
+        var firstAccessFromPostgres: RDD[FirstAccessByDeviceID] = null;
+        try {
+            firstAccessFromPostgres = summary.flatMap{x =>
+                val firstAccessQuery = s"SELECT device_id, first_access FROM $table WHERE device_id = '${x._1.device_id}'"
+                postgresDB.readFirstAccessFromDB(firstAccessQuery)
```
Comment (SanthoshVasabhaktula) [inline platform-modules/batch-models/src/main/scala/org/ekstep/analytics/model/DeviceSummaryModel.scala]: Is this not executed on workers? Does it open a new connection for each request? Why can't we bulk get the data?

File: platform-modules/batch-models/src/main/scala/org/ekstep/analytics/model/DeviceSummaryModel.scala [hunk 3]
```diff
@@ -65,17 +67,33 @@ object DeviceSummaryModel extends IBatchModelTemplate[String, DeviceInput, Devic
             val dial_success = dialcodes_events.filter(f => f.edata.size > 0).length
             val dial_failure = dialcodes_events.filter(f => f.edata.size == 0).length
             val content_downloads = raw.filter(f => "INTERACT".equals(f.eid)).length
-            (index, DeviceSummary(index.device_id, index.channel, CommonUtil.roundDouble(total_ts, 2), total_launches, contents_played, unique_contents_played, content_downloads, DialStats(dial_count, dial_success, dial_failure), DtRange(startTimestamp, endTimestamp), syncts, startTimestamp))
+            (index, DeviceSummary(index.device_id, index.channel, CommonUtil.roundDouble(total_ts, 2), total_launches, contents_played, unique_contents_played, content_downloads, DialStats(dial_count, dial_success, dial_failure), DtRange(startTimestamp, endTimestamp), syncts, new Timestamp(startTimestamp)))
+        }
+        var firstAccessFromPostgres: RDD[FirstAccessByDeviceID] = null;
+        try {
+            firstAccessFromPostgres = summary.flatMap{x =>
+                val firstAccessQuery = s"SELECT device_id, first_access FROM $table WHERE device_id = '${x._1.device_id}'"
```
Comment (SanthoshVasabhaktula) [inline platform-modules/batch-models/src/main/scala/org/ekstep/analytics/model/DeviceSummaryModel.scala]: I think select query can be avoided completely

File: platform-modules/batch-models/src/main/scala/org/ekstep/analytics/updater/UpdateDeviceProfileDB.scala [hunk 4]
```diff
@@ -1,57 +1,96 @@
...
+  val table = AppConf.getConfig("postgres.device.table_name")
+
+  override def preProcess(data: RDD[DerivedEvent], config: Map[String, AnyRef])(implicit sc: SparkContext, fc: FrameworkContext): RDD[DeviceProfileInput] = {
+
+    val filteredEvents = DataFilter.filter(data, Filter("eid", "EQ", Option("ME_DEVICE_SUMMARY")))
+    val newGroupedEvents = filteredEvents.map { event =>
+      (DeviceProfileKey(event.dimensions.did.get), Buffer(event))
+    }.partitionBy(new HashPartitioner(JobContext.parallelization)).reduceByKey((a, b) => a ++ b);
+    val query = s"SELECT * FROM $table;"
+    val testData = postgresDB.readDBData(query)
```
Comment (SanthoshVasabhaktula) [inline platform-modules/batch-models/src/main/scala/org/ekstep/analytics/updater/UpdateDeviceProfileDB.scala]: What is the meaning of test data here

File: platform-modules/batch-models/src/main/scala/org/ekstep/analytics/updater/UpdateDeviceProfileDB.scala [hunk 5]
```diff
@@ -1,57 +1,96 @@
...
-        sc.makeRDD(List(Empty()));
+        val columns = resultMap.keySet.mkString(",")
+        val values = resultMap.values.mkString("','")
+        val queryReq = s"""INSERT INTO $table ($columns) VALUES ('$values') ON CONFLICT (device_id) DO UPDATE SET ($columns) = ('$values')"""
+        postgresDB.insertDataToPostgresDB(queryReq)
+      }
+    }
+    }
+    finally {
+      postgresDB.closeConnection
```
Comment (SanthoshVasabhaktula) [inline platform-modules/batch-models/src/main/scala/org/ekstep/analytics/updater/UpdateDeviceProfileDB.scala]: This is not right - opening and closing a connection should be in each partition

File: platform-modules/batch-models/src/main/scala/org/ekstep/analytics/model/DeviceSummaryModel.scala [hunk 6]
```diff
@@ -65,22 +78,32 @@ object DeviceSummaryModel extends IBatchModelTemplate[String, DeviceInput, Devic
...
+        val summaryData = summary.map(f => (f._1.device_id, f._2))
+        summaryData.leftOuterJoin(postgresData)
+          .map{f =>
+              val defaultDate = CommonUtil.getTimestampFromEpoch(0L)
+          val firstAccessVal = f._2._2.getOrElse(Option(defaultDate))
+          if(firstAccessVal != defaultDate) {
+              f._2._1.copy(firstAccess = firstAccessVal.getOrElse(f._2._1.firstAccess))
+          }
+          else
+              f._2._1
```
Comment (SanthoshVasabhaktula) [inline platform-modules/batch-models/src/main/scala/org/ekstep/analytics/model/DeviceSummaryModel.scala]: This logic doesn't seem right. Here are the scenarios:
1. if firstAccess in db is null - update with the value from the summary event
2. if firstAccess in db is not null - compare this with the value from summary event and use the lowest date.

I don't seem to see the second logic in place

File: platform-modules/batch-models/src/main/scala/org/ekstep/analytics/updater/UpdateDeviceProfileDB.scala [hunk 7]
```diff
@@ -1,57 +1,90 @@
...
+      val avg_ts = if (total_launches == 0) total_ts else CommonUtil.roundDouble(total_ts / total_launches, 2)
+      DeviceProfileOutput(events.index.device_id, Option(first_access), Option(last_access), Option(total_ts), Option(total_launches), Option(avg_ts), prevProfileData.device_spec, prevProfileData.uaspec, prevProfileData.state, prevProfileData.city, prevProfileData.country, prevProfileData.country_code, prevProfileData.state_code, prevProfileData.state_custom, prevProfileData.state_code_custom, prevProfileData.district_custom, prevProfileData.fcm_token, prevProfileData.producer_id, prevProfileData.user_declared_state, prevProfileData.user_declared_district, prevProfileData.api_last_updated_on)
     }
+  }
+
+  override def postProcess(data: RDD[DeviceProfileOutput], config: Map[String, AnyRef])(implicit sc: SparkContext, fc: FrameworkContext): RDD[Empty] = {
+    implicit val sqlContext = new SQLContext(sc)
+    import sqlContext.implicits._
+    val saveMode = SaveMode.Overwrite
+    data.toDF().show()
```
Comment (SanthoshVasabhaktula) [inline platform-modules/batch-models/src/main/scala/org/ekstep/analytics/updater/UpdateDeviceProfileDB.scala]: Remove the debug log

File: platform-modules/batch-models/src/main/scala/org/ekstep/analytics/updater/UpdateDeviceProfileDB.scala [hunk 8]
```diff
@@ -1,57 +1,90 @@
...
+      val total_launches = if (prevProfileData.total_launches.isEmpty) current_launches else current_launches + prevProfileData.total_launches.get
+      val avg_ts = if (total_launches == 0) total_ts else CommonUtil.roundDouble(total_ts / total_launches, 2)
+      DeviceProfileOutput(events.index.device_id, Option(first_access), Option(last_access), Option(total_ts), Option(total_launches), Option(avg_ts), prevProfileData.device_spec, prevProfileData.uaspec, prevProfileData.state, prevProfileData.city, prevProfileData.country, prevProfileData.country_code, prevProfileData.state_code, prevProfileData.state_custom, prevProfileData.state_code_custom, prevProfileData.district_custom, prevProfileData.fcm_token, prevProfileData.producer_id, prevProfileData.user_declared_state, prevProfileData.user_declared_district, prevProfileData.api_last_updated_on)
     }
+  }
+
+  override def postProcess(data: RDD[DeviceProfileOutput], config: Map[String, AnyRef])(implicit sc: SparkContext, fc: FrameworkContext): RDD[Empty] = {
+    implicit val sqlContext = new SQLContext(sc)
+    import sqlContext.implicits._
+    val saveMode = SaveMode.Overwrite
```
Comment (SanthoshVasabhaktula) [inline platform-modules/batch-models/src/main/scala/org/ekstep/analytics/updater/UpdateDeviceProfileDB.scala]: unnecessary declaration

File: platform-modules/batch-models/src/main/scala/org/ekstep/analytics/updater/UpdateDeviceProfileDB.scala [hunk 9]
```diff
@@ -1,57 +1,90 @@
...
+
+  val db = AppConf.getConfig("postgres.db")
+  val url = AppConf.getConfig("postgres.url") + s"$db"
+  val user = AppConf.getConfig("postgres.user")
+  val pass = AppConf.getConfig("postgres.pass")
+
+  val connProperties = new Properties()
+  connProperties.setProperty("Driver", "org.postgresql.Driver")
+  connProperties.setProperty("user", user)
+  connProperties.setProperty("password", pass)
```
Comment (SanthoshVasabhaktula) [inline platform-modules/batch-models/src/main/scala/org/ekstep/analytics/updater/UpdateDeviceProfileDB.scala]: Move this to a common util code as this is duplicated across two data products. may be `CommonUtil.getPostgresConnectionProps()`

File: platform-modules/batch-models/src/main/scala/org/ekstep/analytics/updater/UpdateDeviceProfileDB.scala [hunk 10]
```diff
@@ -1,57 +1,90 @@
...
-        deviceData.map { x => DeviceProfileInput(x._1, x._2._1, x._2._2) }
-    }
+  val className = "org.ekstep.analytics.model.UpdateDeviceProfileDB"
+
+  override def name: String = "UpdateDeviceProfileDB"
+
+  val db = AppConf.getConfig("postgres.db")
+  val url = AppConf.getConfig("postgres.url") + s"$db"
+  val user = AppConf.getConfig("postgres.user")
+  val pass = AppConf.getConfig("postgres.pass")
```
Comment (SanthoshVasabhaktula) [inline platform-modules/batch-models/src/main/scala/org/ekstep/analytics/updater/UpdateDeviceProfileDB.scala]: Move to common util class

File: platform-modules/batch-models/src/test/scala/org/ekstep/analytics/job/summarizer/TestDeviceSummarizer.scala [hunk 11]
```diff
@@ -1,14 +1,52 @@
...
+               |    state TEXT,
+               |    state_code TEXT,
+               |    state_code_custom TEXT,
+               |    state_custom TEXT,
+               |    total_launches bigint,
+               |    total_ts float,
+               |    uaspec json,
+               |    updated_date TIMESTAMP,
+               |    user_declared_district TEXT,
+               |    user_declared_state TEXT)""".stripMargin)
```
Comment (SanthoshVasabhaktula) [inline platform-modules/batch-models/src/test/scala/org/ekstep/analytics/job/summarizer/TestDeviceSummarizer.scala]: Where is the user_declared_on field?

File: platform-modules/batch-models/src/main/scala/org/ekstep/analytics/model/DeviceSummaryModel.scala [hunk 12]
```diff
@@ -1,28 +1,40 @@
...
+    val db = AppConf.getConfig("postgres.db")
+    val url = AppConf.getConfig("postgres.url") + s"$db"
+    val user = AppConf.getConfig("postgres.user")
+    val pass = AppConf.getConfig("postgres.pass")
+
+    val connProperties = new Properties()
+    connProperties.setProperty("Driver", "org.postgresql.Driver")
+    connProperties.setProperty("user", user)
+    connProperties.setProperty("password", pass)
+
```
Comment (SanthoshVasabhaktula) [inline platform-modules/batch-models/src/main/scala/org/ekstep/analytics/model/DeviceSummaryModel.scala]: Move to common util

File: platform-modules/batch-models/src/main/scala/org/ekstep/analytics/model/DeviceSummaryModel.scala [hunk 13]
```diff
@@ -65,22 +78,32 @@ object DeviceSummaryModel extends IBatchModelTemplate[String, DeviceInput, Devic
...
-        summary.leftOuterJoin(firstAccessFromCassandra)
-          .map{ x =>
-              val firstAccessValue = x._2._2.getOrElse(Option(0L))
-              if(firstAccessValue.getOrElse(0L) != 0)
-                  x._2._1.copy(firstAccess = firstAccessValue.getOrElse(x._2._1.firstAccess))
-              else x._2._1
+
+        implicit val sqlContext = new SQLContext(sc)
+        val responseDf = sqlContext.sparkSession.read.jdbc(url, Constants.DEVICE_PROFILE_TABLE, connProperties).select("device_id","first_access")
+        responseDf.show()
```
Comment (SanthoshVasabhaktula) [inline platform-modules/batch-models/src/main/scala/org/ekstep/analytics/model/DeviceSummaryModel.scala]: remove debug log

File: platform-modules/batch-models/src/main/scala/org/ekstep/analytics/updater/UpdateDeviceProfileDB.scala [hunk 14]
```diff
@@ -1,57 +1,90 @@
...
+    implicit val sqlContext = new SQLContext(sc)
+
+    val filteredEvents = DataFilter.filter(data, Filter("eid", "EQ", Option("ME_DEVICE_SUMMARY")))
+
+    val newGroupedEvents = filteredEvents.map { event =>
+      (DeviceProfileKey(event.dimensions.did.get), Buffer(event))
+    }.partitionBy(new HashPartitioner(JobContext.parallelization)).reduceByKey((a, b) => a ++ b)
+
+    val resDf = sqlContext.sparkSession.read.jdbc(url, Constants.DEVICE_PROFILE_TABLE, connProperties)
+    resDf.show()
```
Comment (SanthoshVasabhaktula) [inline platform-modules/batch-models/src/main/scala/org/ekstep/analytics/updater/UpdateDeviceProfileDB.scala]: remove debug log

File: platform-modules/batch-models/src/test/scala/org/ekstep/analytics/model/TestDeviceSummaryModel.scala [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+         |    state TEXT,
+         |    state_code TEXT,
+         |    state_code_custom TEXT,
+         |    state_custom TEXT,
+         |    total_launches bigint,
+         |    total_ts float,
+         |    uaspec json,
+         |    updated_date TIMESTAMP,
+         |    user_declared_district TEXT,
+         |    user_declared_state TEXT)""".stripMargin)
```
Comment (SanthoshVasabhaktula) [inline platform-modules/batch-models/src/test/scala/org/ekstep/analytics/model/TestDeviceSummaryModel.scala]: user_declared_on field is missing

File: platform-modules/batch-models/src/main/scala/org/ekstep/analytics/model/DeviceSummaryModel.scala (hunk 6 above)
Comment (utk14) [reply platform-modules/batch-models/src/main/scala/org/ekstep/analytics/model/DeviceSummaryModel.scala]: It was not in the previous implementation. I have updated the changes

**Pull request merged**